Ansys product compatibility version documentation (#170)
pyansys/dev-guide: doc/source/how-to/compatibility.rst

> Ansys products compatibility

Reviewer: ```suggestion
Ansys product compatibility
```

============================

> As the different PyAnsys libraries evolve, backwards and forwards compatibility

Reviewer: ```suggestion
As the different PyAnsys libraries evolve, backward and forward compatibility
```

Reviewer: Google style guide says no "s".

> issues may rise. Some of the most common cases are:

Reviewer: ```suggestion
issues can occur. Some of the most common cases are:
```

* An Ansys product has implemented certain features in its new server version

>   which are not available in previous server versions. This causes backwards

Reviewer: ```suggestion
  that are not available in previous server versions. This causes backward
```

  incompatibility issues.
* New server versions have suppressed support for certain operations after a

>   a given version. This causes forwards incompatibility issues.

Reviewer: ```suggestion
  a given version. This causes forward incompatibility issues.
```

Though there are different ways to handle these issues, some of the PyAnsys libraries,
such as `PyMAPDL <https://github.com/pyansys/pymapdl>`_ and

> `PyDPF-Core <https://github.com/pyansys/pydpf-core>`_, have started to handle them in

Reviewer: ```suggestion
`PyDPF-Core <https://github.com/pyansys/pydpf-core>`_, handle them in
```

> the same way. In order to homogenize implementations throughout the libraries,
> it is recommended to follow their approach.

Reviewer: ```suggestion
the same way. To homogenize implementations in PyAnsys libraries,
following their approach is recommended.
```

``check_version.py`` module approach
------------------------------------

> As mentioned before, PyAnsys libraries have started to implement *version checking*
> modules which allow to determine whether the Ansys product server you are connecting
> to provides support for certain operations. Examples of this module may be found for
> the following libraries:

Reviewer: ```suggestion
*Version checking* modules determine whether the Ansys product server you are connecting
to provides support for certain operations. See the ``check_version.py`` files for the following PyAnsys
libraries for implementation examples:
```

* `PyMAPDL check_version.py module <https://github.com/pyansys/pymapdl/blob/main/src/ansys/mapdl/core/check_version.py>`_
* `PyDPF-Core check_version.py module <https://github.com/pyansys/pydpf-core/blob/master/ansys/dpf/core/check_version.py>`_

One of the easiest ways to keep track of the versions supported is setting up a

> **minimum version** data structure, in which forwards compatibility is ensured.

Reviewer: ```suggestion
**minimum version** data structure, in which forward compatibility is ensured.
```

> Server versions prior to the minimum version do not have access to this

Reviewer: ```suggestion
Server versions  earlier than the minimum version do not have access to this
```

> functionality. In the case of the `PyMAPDL check_version.py module`_, this is the

Reviewer: Is the link to a PY file? If so, use "In the case of the ``PyMAPDL check_version.py`` file, the ``VERSIONS_MAP`` structure provides the minimum and maximum data structure."

I wasn't sure what "this is" referred to in your original sentence.

Reviewer 2: ```suggestion
functionality. In the case of `ansys/mapdl/core/check_version.py`_, this is the
```

``VERSIONS_MAP`` structure.

> Most Ansys products provide forwards compatibility, meaning that whatever feature

Reviewer: ```suggestion
Most Ansys products provide forward compatibility, meaning that whatever feature
```

introduced in a previous version, is typically supported in upcoming versions
as well. In case that a feature was decided to be suppressed, this would lead to
backwards compatibility issues. The same would happen with the PyAnsys servers wrapping

> the Ansys products. Thus, for this cases, it would also be necessary to create a

Reviewer 3: ```suggestion
the Ansys products. Thus, for this case, it would also be necessary to create a
```

> Most Ansys products provide forwards compatibility, meaning that whatever feature
> introduced in a previous version, is typically supported in upcoming versions
> as well. In case that a feature was decided to be suppressed, this would lead to
> backwards compatibility issues. The same would happen with the PyAnsys servers wrapping
> the Ansys products. Thus, for this case, it would also be necessary to create a
> similar data structure, but in this case for a **maximum version**. Though there are
> no implementations yet of this feature, the way it should work would be analogous to
> the way the **minimum version** mechanism does.

Reviewer: ```suggestion
Most Ansys products provide forward compatibility, meaning that features
introduced in an earlier version are also supported in later versions. Suppressing
a feature would lead to a backward compatibility issue.

The same happen with the PyAnsys servers wrapping
the Ansys products. Thus, for this case, it would also be necessary to create a
similar data structure, but in this case for a **maximum version**. Though there are
no implementations yet of this feature, the way it should work would be analogous to
the way the **minimum version** mechanism does.
```

``version_requires`` decorator
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> How this logic is applied to the different functionalities and methods available in
> the client is by using the ``@version_requires`` decorator. The PyAnsys libraries define
> this decorator as seen in `PyMAPDL check_version.py module`_ or
> `PyDPF-Core check_version.py module`_ and it is typically used as follows:

Reviewer: ```suggestion
The ``@version_requires`` decorator applies version logic to the different
functionalities and methods available in the client. You can see how this
decorator is used in the `PyMAPDL check_version.py module`_ and
`PyDPF-Core check_version.py module`_. Here is a general example:

```

Reviewer: Again, I think the link should be the name of the file with the word file following it.

.. code:: python

    class Client():
        def __init__(self):

>             '''Connects to a fake server'''

Reviewer 2: These should be `"""`

>             '''Connects to a fake server'''

Reviewer: ```suggestion
            '''Connect to a fake server.'''
```

            self._server = FakeServer()

        @version_requires((0, 1, 3))  # require 0.1.3
        def meth_a(self):

>             '''calls method a on the 'server''''

Reviewer: ```suggestion
            '''Call method ``a`` on the server.''''
```

            return self._server.meth_a()

        @version_requires((0, 2, 3))  # require 0.2.3
        def meth_b(self):

>             '''calls method b on the 'server''''

Reviewer: ```suggestion
            '''Call method ``b`` on the server.
```

            return self._server.meth_b()

> That way, whenever a ``Client`` object is created and a call to any of its methods is done,

Reviewer: ```suggestion
Whenever a ``client`` object is created and a call is made to any of its methods,
```

> such as ``meth_a()``, the decorator first checks that the version dealt with is above the one
> specified in the decorator's call. Otherwise, a ``VersionError`` is raised.

Reviewer: ```suggestion
such as ``meth_a()``, the decorator checks that the version is later than the one
specified in the decorator's call. Otherwise, a ``VersionError`` is raised.
```

Reviewer: Must the version be later or can it be the same and later?

> The ``self._server`` member of the class (that is, the server the client is connected to) is
> expected to have information on its version (that is, a ``self._server._server_version`` attribute),
> which the decorator uses to determine if the version is above the threshold or not.

Reviewer: ```suggestion
The ``self._server`` member of the class is the server that the client is connected to. It is
expected to have information on its version in its ``self._server._server_version`` attribute.
The decorator uses this information to determine if the version is above the threshold or not.
```

> This same kind of logic could be implemented for forwards incompatibility, by creating a new
> ``@max_version_supported`` decorator. If this were to happen, it is suggested that the
> ``@version_requires`` decorator is changed to ``@min_version_required``.

Reviewer: ```suggestion
You can create a ``@max_version_supported`` decorator to implement this same
kind of logic for forward incompatibility. You should change the ``@version_requires``
decorator to ``@min_version_required``.
```

Ansys product compatibility
============================
As the different PyAnsys libraries evolve, backward and forward compatibility
issues can occur. Some of the most common cases are:

* An Ansys product has implemented certain features in its new server version
  that are not available in previous server versions. This causes backward
  incompatibility issues.
* New server versions have suppressed support for certain operations after a
  a given version. This causes forward incompatibility issues.

Though there are different ways to handle these issues, some of the PyAnsys libraries,
such as `PyMAPDL <https://github.com/pyansys/pymapdl>`_ and
`PyDPF-Core <https://github.com/pyansys/pydpf-core>`_, handle them in
the same way. To homogenize implementations in PyAnsys libraries,
following their approach is recommended.

``check_version.py`` module approach
------------------------------------
A *version checking* module determines whether the Ansys product server you are connecting
to provides support for certain operations. For implementation examples, see the
``check_version.py`` files for the following PyAnsys libraries:

> * `PyMAPDL check_version.py module <https://github.com/pyansys/pymapdl/blob/main/src/ansys/mapdl/core/check_version.py>`_

Reviewer 2: ```suggestion
* PyMAPDL's `check_version.py <https://github.com/pyansys/pymapdl/blob/main/src/ansys/mapdl/core/check_version.py>`_
```

> * `PyMAPDL check_version.py module <https://github.com/pyansys/pymapdl/blob/main/src/ansys/mapdl/core/check_version.py>`_
> * `PyDPF-Core check_version.py module <https://github.com/pyansys/pydpf-core/blob/master/ansys/dpf/core/check_version.py>`_

Reviewer 2: ```suggestion
* `ansys/mapdl/core/check_version.py <https://github.com/pyansys/pymapdl/blob/main/src/ansys/mapdl/core/check_version.py>`_
* `ansys/dpf/core/check_version.py <https://github.com/pyansys/pydpf-core/blob/master/ansys/dpf/core/check_version.py>`_
```
